### ISSUE TITLE
Vertical streak artifacts in MLEM

### ISSUE THREAD

**Issue by rijobro** [NiftyPET/NIPET#20]
I've noticed vertical streak artifacts on my MLEM reconstructions (also those in STIR using the NP projectors). Am I doing something wrong or is this a bug?

I've created a small jupyter notebook ([attached](https://github.com/NiftyPET/NIPET/files/4748232/NP_MLEM.txt), saved as .txt) that demonstrates what I mean. Very similar to @casperdcl's [demo](https://github.com/NiftyPET/NIPET/blob/master/examples/demo.ipynb), but with no cylindrical truncation or PSF. I've used norm and randoms, but no scatter or attenuation.

The important chunk of code is below, and the orthogonal views after 50 iterations is after that.

```
iters = 50
prompts = m['psino'].astype(np.float32)
additive = rands
multiplicative = norm

def make_non_zero(A):
    """Make an array non-zero"""
    A[A <= 0] = np.min(A[A > 0]) * 0.001
    return A

def fwd(im):
    """Forward project"""
    out = nipet.frwd_prj(im, mMRpars)
    if multiplicative is not None:
        out *= multiplicative
    if additive is not None:
        out += additive
    out = make_non_zero(out)
    return out

def bck(sino):
    """Back project"""
    sino_to_proj = np.copy(sino)
    if multiplicative is not None:
        sino_to_proj *= multiplicative
    out = nipet.back_prj(sino_to_proj, mMRpars)
    return out

bck_1s = bck(np.ones_like(prompts))
bck_1s = make_non_zero(bck_1s)
inv_bck_1s = 1 / bck_1s

recon_mlem = [np.ones_like(bck_1s)]
for k in trange(iters, desc="MLEM"):
    fprj = fwd(recon_mlem[-1])
    recon_mlem.append(recon_mlem[-1] * inv_bck_1s * bck(prompts / fprj))
```

![image](https://user-images.githubusercontent.com/33289025/84073490-d1679000-a9c8-11ea-8959-c629dbb062fb.png)

Lastly, there's also some horizontal artifacts, that can be seen on this zoomed image of the chin:

![image](https://user-images.githubusercontent.com/33289025/84073773-5783d680-a9c9-11ea-8d7f-05384ad25b7c.png)

**Comment by rijobro** [comment]
Hi, a little bump on this. Has anyone had a chance to have a look? 
I wonder if this potential bug exists in the forward and back projector, but doesn't exist in the GPU OSEM implementation?

**Comment by pjmark** [comment]
Yes, I come across them particularly when using NAC recons.  These artefacts seem to be caused by the highly defined edges of the bed and some other factors.  When the mu-maps are well aligned this problem should not appear straight to the beholder.  I noticed it even on Siemens reconstructions to some extent as well.

Cheers,
P
________________________________

**Comment by casperdcl** [comment]
@rijobro / @pjmark I can confirm that I can reproduce these images. However if I add attenuation (using the object & hardware attenuation as in the example notebook) as well as scatter correction then I get this (basically the example notebook MLEM without PSF):

![image](https://user-images.githubusercontent.com/10780059/100776741-d0a23a80-33fc-11eb-8a63-1abb840c1bd1.png)

**Comment by KrisThielemans** [comment]
that's interesting. I guess you could try "NAC" with only hardware mu-maps (i.e. without patient) to check if that's it?

Weird thing though is that I don't think we have this with STIR projectors.

**Comment by casperdcl** [comment]
hmm... I've tried a few permutations as @rijobro describes with no scatter nor PSF (click on images for full size):

Hardware AC|Object AC|Both
--|--|--
![image](https://user-images.githubusercontent.com/10780059/100780800-17defa00-3402-11eb-8295-b5bd6f29ee63.png)|![image](https://user-images.githubusercontent.com/10780059/100782193-faab2b00-3403-11eb-8cf0-84d39619ccd2.png)|![image](https://user-images.githubusercontent.com/10780059/100783445-b02aae00-3405-11eb-9d0f-98792efd31d2.png)

Really looks like full AC is required...

Are STIR projectors Siddon-based (infinitely narrow LoRs)? NiftyPET uses Siddon so really should be using ~2.5mm FWHM PSF to approximate a more realistic projector - maybe if STIR does this the NAC issue gets hidden?

**Comment by rijobro** [comment]
I wonder if there's anything we can test with simulated data. For example, what would we expect if we projected the brainweb data and then NAC reconstructed that?

**Comment by pjmark** [comment]
Those would be pretty good.  I have done lots of them for simulation.  But whenever there is discrepancy between attenuation and emission data, you get those artefacts

P
________________________________

**Comment by casperdcl** [comment]
> I wonder if there's anything we can test with simulated data. For example, what would we expect if we projected the brainweb data and then NAC reconstructed that?

could use `nipet.simulate_sino(phantom, mu_map, mMRpars, simulate_d=True, mu_input=True)` and then `poisson()` followed by NAC recon...

**Comment by KrisThielemans** [comment]
> Are STIR projectors Siddon-based (infinitely narrow LoRs)? NiftyPET uses Siddon so really should be using ~2.5mm FWHM PSF to approximate a more realistic projector - maybe if STIR does this the NAC issue gets hidden?

STIR's ray-tracing projector does ray-tracing (!). It happens to do it via Siddon algorithm. However, we rarely use it with just a single ray (although that's still the default). Using 5-10 rays improves discretisation artefacts dramatically.

I guess we could do a STIR NAC with only 1 ray if you'd like to compare.

Normally I'd expect that if you simulate and reconstruct with the same projector, everything works nicely...

**Comment by casperdcl** [comment]
FYI artefacts also visible in NAC OSEM

![image](https://user-images.githubusercontent.com/10780059/101170567-902e0100-3636-11eb-9f04-077fcd901fc6.png)

> I guess we could do a STIR NAC with only 1 ray if you'd like to compare.

yes would be a nice to see

**Comment by rijobro** [comment]
@casperdcl my original reconstruction and the one in your notebook use projectors where the data is copied between the GPU and CPU. But there's also an OSEM implementation that works purely on the GPU. Do you think you'd be able to run an NAC recon using that functionality? I'm sure that the result would be the same if it's down to ray tracing, but I just wanted to confirm.

**Comment by casperdcl** [comment]
@rijobro my last comment (just above yours) *is* the NAC OSEM you refer to...

**Comment by rijobro** [comment]
ah. well good to tick that one off the list at least!

**Comment by rijobro** [comment]
Just reconstructed the same data using STIR/SIRF. Both NACs with norm and randoms, top row is 1 ray, bottom row is 10 rays.

![image](https://user-images.githubusercontent.com/33289025/102195610-b5cccd00-3eb6-11eb-81ba-0b62c1bfdaff.png)

**Comment by KrisThielemans** [comment]
interesting... I don't think I've seen this with other scanners, so should be a feature of the gaps that make it worse.

@rijobro what about trying with 100 LORs. shouldn't take too much longer (weirdly enough).

I have no clue though why it would disappear with full AC only. There's obviously going to be similar discretisation artefacts in the AC factors and it's marginally conceivable that they cancel out a little bit, but I'd have expected it then to mostly disappear in the "object AC" recon.

Still, there might be a good thing here: Richard might have wrapped NiftyPET's projectors correctly into STIR!

**Comment by pjmark** [comment]
This is a tough one.  But there are clues here and there.  For example, when the bed is removed they never appear.  Look at this 24hr scan of Ge68 cylindrical phantom reconstructed without AC:

![axial_ge68_on_mMR](https://user-images.githubusercontent.com/6179486/102288233-fd456e80-3f33-11eb-8c49-0bd36e4fc168.png)

It looks pure and nice.

**Comment by KrisThielemans** [comment]
interesting. As this cylinder with head-coil? Likely not.

I can see that the bed/coil introduces high-frequency AC artefacts (and the coil symmetric patterns). I can also see that if you don't do pre-smoothing of your mu-map that you have resolution mismatch with the AC, and therefore have remaining artefacts when you do "hardware AC only", but it's a mystery to me why they then would disappear when adding in the object AC.

@pjmark would you be able to make the cylinder data available? Ideally we'd reconstruct it with STIR (with both its own projector and the NiftyPET-one) as well.

**Comment by rijobro** [comment]
1, 10, 100 rays:
![image](https://user-images.githubusercontent.com/33289025/102342830-b5553480-3f91-11eb-8cd7-a8b961f0d72f.png)

**Comment by KrisThielemans** [comment]
ok. nice. so it doesn't go away, but reduces dramatically.

**Comment by ashgillman** [comment]
One thing I can note in the 1-ray reconstructions is that the artefacts appear to be rotationally symmetrical around the gantry centre, and that the repetition angle seems to be either 90 or 180 degrees.

To me they look a bit like hyperbolas

![image](https://user-images.githubusercontent.com/816777/102432845-2c0c3380-405f-11eb-8d73-fe04e26a04e7.png)

I don't know the significance of the shape.. But the symmetries about x and y also point towards VOI/LOI errors with the ray tracing. And they're somehow exacerbated by poor AC (or rather somehow compensated with good AC maybe?)

But I think you already worked that out...

**Comment by pjmark** [comment]
> interesting. As this cylinder with head-coil? Likely not.
> 
> I can see that the bed/coil introduces high-frequency AC artefacts (and the coil symmetric patterns). I can also see that if you don't do pre-smoothing of your mu-map that you have resolution mismatch with the AC, and therefore have remaining artefacts when you do "hardware AC only", but it's a mystery to me why they then would disappear when adding in the object AC.
> 
> @pjmark would you be able to make the cylinder data available? Ideally we'd reconstruct it with STIR (with both its own projector and the NiftyPET-one) as well.

Yes, but the data size is >100GB so difficult to share.  Any ideas?

**Comment by KrisThielemans** [comment]
> the data size is >100GB so difficult to share. Any ideas?

It looks like [OneDrive can do 100GB per file](https://support.microsoft.com/en-us/office/invalid-file-names-and-file-types-in-onedrive-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa?ui=en-us&rs=en-us&ad=us)  these days. Nevertheless probably best to split. see e.g. https://www.baeldung.com/linux/split-files
